src/Features/LanguageServer/Protocol/Handler/Diagnostics/AbstractWorkspaceDiagnosticSourceProvider.cs

```diff
@@ -0,0 +1,207 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.EditAndContinue;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.SolutionCrawler;
+using Microsoft.CodeAnalysis.TaskList;
+using Roslyn.Utilities;
+
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler.Diagnostics;
+
+internal abstract class AbstractWorkspaceDiagnosticSourceProvider(
+    IDiagnosticAnalyzerService diagnosticAnalyzerService,
+    IGlobalOptionService globalOptions,
+    ImmutableArray<string> sourceNames)
+    : IDiagnosticSourceProvider
+{
+    public bool IsDocument => false;
+    public ImmutableArray<string> SourceNames => sourceNames;
+
+    public async ValueTask<ImmutableArray<IDiagnosticSource>> CreateDiagnosticSourcesAsync(RequestContext context, string sourceName, CancellationToken cancellationToken)
+    {
+        // If we're being called from razor, we do not support WorkspaceDiagnostics at all.  For razor, workspace
+        // diagnostics will be handled by razor itself, which will operate by calling into Roslyn and asking for
+        // document-diagnostics instead.
+        if (context.ServerKind == WellKnownLspServerKinds.RazorLspServer)
+            return [];
+
+        if (sourceName == PullDiagnosticCategories.Task)
+            return GetTaskListDiagnosticSources(context, globalOptions);
+
+        if (sourceName == PullDiagnosticCategories.EditAndContinue)
```

Comment:
this feels a little weird to me - instead of this type checking the source name and creating different kinds of sources, it seems like each of the possible sources here should be in a separate provider
e.g.
1. a tasklist source provider
2. an enc source provider
3. workspace diags source provider
4. a source for each of the various analyzer/compiler diags (could share a common base since they're so similar)

then the manager is responsible for doing these checks (it gets passed in the source name)